File: backend/app/services/question_service.py
import json
import math
from typing import List

from fastapi import HTTPException, status
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models import Bookmark, Option, Question, UserProgress
from app.schemas.question import (
    OptionResponse,
    OptionWithAnswerResponse,
    QuestionListItem,
    QuestionPage,
    QuestionResponse,
    SubmitAnswerResponse,
    UserProgressBrief,
)


async def get_questions_page(
    db: AsyncSession, exam_id: int, user_id: int, page: int = 1, per_page: int = 50
) -> QuestionPage:
    per_page = min(per_page, 200)  # Cap to prevent abuse
    count_result = await db.execute(
        select(func.count()).select_from(Question).where(Question.exam_id == exam_id)
    )
    total = count_result.scalar() or 0
    total_pages = math.ceil(total / per_page) if total > 0 else 1

    offset = (page - 1) * per_page
    result = await db.execute(
        select(Question)
        .where(Question.exam_id == exam_id)
        .order_by(Question.order_index)
        .offset(offset)
        .limit(per_page)
    )
    questions = result.scalars().all()
    question_ids = [q.id for q in questions]

    progress_result = await db.execute(
        select(UserProgress)
        .where(UserProgress.user_id == user_id, UserProgress.question_id.in_(question_ids))
    )
    progress_map = {p.question_id: p for p in progress_result.scalars().all()}

    bookmark_result = await db.execute(
        select(Bookmark.question_id)
        .where(Bookmark.user_id == user_id, Bookmark.question_id.in_(question_ids))
    )
    bookmarked_ids = {row[0] for row in bookmark_result.all()}

    items = []
    for q in questions:
        p = progress_map.get(q.id)
        items.append(
            QuestionListItem(
                id=q.id,
                external_id=q.external_id,
                question_type=q.question_type,
                num_correct=q.num_correct,
                order_index=q.order_index,
                is_attempted=p is not None,
                is_correct=p.is_correct if p else None,
                is_bookmarked=q.id in bookmarked_ids,
            )
        )

    return QuestionPage(items=items, total=total, page=page, per_page=per_page, total_pages=total_pages)


async def get_question_detail(db: AsyncSession, question_id: int, user_id: int) -> QuestionResponse:
    result = await db.execute(
        select(Question)
        .options(selectinload(Question.options))
        .where(Question.id == question_id)
    )
    question = result.scalar_one_or_none()
    if not question:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Question not found")

    bm_result = await db.execute(
        select(Bookmark).where(Bookmark.user_id == user_id, Bookmark.question_id == question_id)
    )
    is_bookmarked = bm_result.scalar_one_or_none() is not None

    prog_result = await db.execute(
        select(UserProgress).where(UserProgress.user_id == user_id, UserProgress.question_id == question_id)
    )
    progress = prog_result.scalar_one_or_none()
    user_progress = None
    if progress:
        user_progress = UserProgressBrief(
            is_correct=progress.is_correct,
            selected_option_ids=json.loads(progress.selected_option_ids),
        )

    options = [
        OptionResponse(id=o.id, label=o.label, option_text=o.option_text)
        for o in question.options
    ]

    return QuestionResponse(
        id=question.id,
        external_id=question.external_id,
        question_text=question.question_text,
        question_type=question.question_type,
        num_correct=question.num_correct,
        order_index=question.order_index,
        options=options,
        is_bookmarked=is_bookmarked,
        user_progress=user_progress,
    )


async def submit_answer(
    db: AsyncSession, question_id: int, user_id: int, selected_option_ids: List[int]
) -> SubmitAnswerResponse:
    result = await db.execute(
        select(Question)
        .options(selectinload(Question.options))
        .where(Question.id == question_id)
    )
    question = result.scalar_one_or_none()
    if not question:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Question not found")

    correct_ids = {o.id for o in question.options if o.is_correct}
    is_correct = set(selected_option_ids) == correct_ids

    prog_result = await db.execute(
        select(UserProgress).where(
            UserProgress.user_id == user_id, UserProgress.question_id == question_id
        )
    )
    progress = prog_result.scalar_one_or_none()
    if progress:
        progress.is_correct = is_correct
        progress.selected_option_ids = json.dumps(selected_option_ids)
    else:
        progress = UserProgress(
            user_id=user_id,
            question_id=question_id,
            is_correct=is_correct,
            selected_option_ids=json.dumps(selected_option_ids),
        )
        db.add(progress)

    await db.commit()

    options_with_answers = [
        OptionWithAnswerResponse(
            id=o.id, label=o.label, option_text=o.option_text, is_correct=o.is_correct
        )
        for o in question.options
    ]

    return SubmitAnswerResponse(
        is_correct=is_correct,
        correct_option_ids=list(correct_ids),
        explanation=question.explanation,
        options=options_with_answers,
    )

File: workers-api/src/services/question-service.ts
import { eq, and, count, sql, inArray } from "drizzle-orm";
import type { Database } from "../db/client";
import { questions, options, userProgress, bookmarks } from "../db/schema";
import { AppError } from "../lib/errors";

// D1 limits bound parameters to 100 per query.
// Batch inArray queries to stay under the limit.
const D1_PARAM_BATCH = 95;

async function batchedQuery<T>(
  ids: number[],
  queryFn: (chunk: number[]) => Promise<T[]>
): Promise<T[]> {
  if (ids.length <= D1_PARAM_BATCH) return queryFn(ids);
  const results: T[] = [];
  for (let i = 0; i < ids.length; i += D1_PARAM_BATCH) {
    results.push(...(await queryFn(ids.slice(i, i + D1_PARAM_BATCH))));
  }
  return results;
}

export async function getQuestionsPage(
  db: Database,
  examId: number,
  userId: number,
  page: number,
  perPage: number
) {
  perPage = Math.min(perPage, 200);
  const offset = (page - 1) * perPage;

  const [totalResult] = await db
    .select({ count: count() })
    .from(questions)
    .where(eq(questions.examId, examId));

  const total = totalResult.count;
  const totalPages = Math.ceil(total / perPage);

  const questionRows = await db
    .select({
      id: questions.id,
      externalId: questions.externalId,
      questionType: questions.questionType,
      numCorrect: questions.numCorrect,
      orderIndex: questions.orderIndex,
    })
    .from(questions)
    .where(eq(questions.examId, examId))
    .orderBy(questions.orderIndex)
    .limit(perPage)
    .offset(offset);

  if (questionRows.length === 0) {
    return { items: [], total, page, per_page: perPage, total_pages: totalPages };
  }

  const qIds = questionRows.map((q) => q.id);

  const progressRows = await batchedQuery(qIds, (chunk) =>
    db
      .select({
        questionId: userProgress.questionId,
        isCorrect: userProgress.isCorrect,
      })
      .from(userProgress)
      .where(and(eq(userProgress.userId, userId), inArray(userProgress.questionId, chunk)))
  );

  const progressMap = new Map(progressRows.map((p) => [p.questionId, p.isCorrect]));

  const bookmarkRows = await batchedQuery(qIds, (chunk) =>
    db
      .select({ questionId: bookmarks.questionId })
      .from(bookmarks)
      .where(and(eq(bookmarks.userId, userId), inArray(bookmarks.questionId, chunk)))
  );

  const bookmarkSet = new Set(bookmarkRows.map((b) => b.questionId));

  const items = questionRows.map((q) => ({
    id: q.id,
    external_id: q.externalId,
    question_type: q.questionType,
    num_correct: q.numCorrect,
    order_index: q.orderIndex,
    is_attempted: progressMap.has(q.id),
    is_correct: progressMap.get(q.id) ?? null,
    is_bookmarked: bookmarkSet.has(q.id),
  }));

  return { items, total, page, per_page: perPage, total_pages: totalPages };
}

export async function getQuestionDetail(
  db: Database,
  questionId: number,
  userId: number
) {
  const question = await db.query.questions.findFirst({
    where: eq(questions.id, questionId),
  });

  if (!question) throw new AppError(404, "Question not found");

  const opts = await db
    .select()
    .from(options)
    .where(eq(options.questionId, questionId))
    .orderBy(options.orderIndex);

  const progress = await db.query.userProgress.findFirst({
    where: and(eq(userProgress.userId, userId), eq(userProgress.questionId, questionId)),
  });

  const bookmark = await db.query.bookmarks.findFirst({
    where: and(eq(bookmarks.userId, userId), eq(bookmarks.questionId, questionId)),
  });

  return {
    id: question.id,
    external_id: question.externalId,
    question_text: question.questionText,
    question_type: question.questionType,
    num_correct: question.numCorrect,
    order_index: question.orderIndex,
    options: opts.map((o) => ({
      id: o.id,
      label: o.label,
      option_text: o.optionText,
    })),
    is_bookmarked: !!bookmark,
    user_progress: progress
      ? {
          is_correct: progress.isCorrect,
          selected_option_ids: JSON.parse(progress.selectedOptionIds) as number[],
        }
      : null,
  };
}

export async function submitAnswer(
  db: Database,
  questionId: number,
  userId: number,
  selectedOptionIds: number[]
) {
  const question = await db.query.questions.findFirst({
    where: eq(questions.id, questionId),
  });
  if (!question) throw new AppError(404, "Question not found");

  const opts = await db
    .select()
    .from(options)
    .where(eq(options.questionId, questionId))
    .orderBy(options.orderIndex);

  const correctIds = new Set(opts.filter((o) => o.isCorrect).map((o) => o.id));
  const selectedSet = new Set(selectedOptionIds);
  const isCorrect =
    correctIds.size === selectedSet.size && [...correctIds].every((id) => selectedSet.has(id));

  // Upsert user progress
  await db
    .insert(userProgress)
    .values({
      userId,
      questionId,
      isCorrect,
      selectedOptionIds: JSON.stringify(selectedOptionIds),
    })
    .onConflictDoUpdate({
      target: [userProgress.userId, userProgress.questionId],
      set: {
        isCorrect,
        selectedOptionIds: JSON.stringify(selectedOptionIds),
        attemptedAt: sql`datetime('now')`,
      },
    });

  return {
    is_correct: isCorrect,
    correct_option_ids: [...correctIds],
    explanation: question.explanation,
    options: opts.map((o) => ({
      id: o.id,
      label: o.label,
      option_text: o.optionText,
      is_correct: o.isCorrect,
    })),
  };
}
